metagraph/src/annotation/taxonomy/taxonomic_db.hpp

```diff
@@ -0,0 +1,167 @@
+     * @param [input] reversed_lookup_table maps taxid (before normalizing) to accession version.
+     * @param [output] tree -> tree stored as list of children.
+     * @param [output] root_node -> normalized id of the root of the tree.
+     */
+    void read_tree(const std::string &taxo_tree_filepath,
+                   const tsl::hopscotch_map<TaxId, AccessionVersion> &reversed_lookup_table,
+                   ChildrenList &tree, NormalizedTaxId &root_node);
+
+    /**
+     * Reads the lookup table and returns the reversed lookup table for the relevant accession versions.
+     *
+     * @param [input] lookup_table_filepath path to a ".accession2taxid" file.
+     * @param [input] input_accessions set containing all the accession version in the input.
+     * @param [output] reversed_lookup_table maps taxid (before normalizing) to accession version.
+     */
+    void read_lookup_table(const std::string &lookup_table_filepath,
+                           const tsl::hopscotch_set<AccessionVersion> &input_accessions,
+                           tsl::hopscotch_map<TaxId, AccessionVersion> &reversed_lookup_table);
+
+    /**
+     * Reads the fasta headers and returns all the relevant accession versions.
+     *
+     * @param [input] fasta_headers_filepath path to a ".fasta.fai" file.
+     * @param [output] input_accessions set containing all the accession version in the fasta input.
+     */
+    void get_input_accessions(const std::string &fasta_headers_filepath,
+                              tsl::hopscotch_set<AccessionVersion> &input_accessions);
+
+    /**
+     * Computes the rmq data in "this->rmq_data". Beside this, calculates
+     *      the fast tables: 'this->precalc_log' and 'this->precalc_pow2'.
+     *
+     * @param [input] tree_linearization -> the linearization of the received tree.
+     */
+    void rmq_preprocessing(const std::vector<NormalizedTaxId> &tree_linearization);
+
+    /**
+     * dfs_statistics calculates a tree_linearization, "this->node_depth" and
+     *      'this->node_to_linearization_idx'.
+     *
+     * @param [input] node - the node that is currently processed.
+     * @param [input] tree -> tree stored as list of children.
+     * @param [output] tree_linearization -> the linearization of the received tree.
+     */
+    void dfs_statistics(const NormalizedTaxId &node, const ChildrenList &tree,
+                        std::vector<NormalizedTaxId> &tree_linearization);
+
+    NormalizedTaxId find_lca(const NormalizedTaxId &taxid1, const NormalizedTaxId &taxid2);
+    NormalizedTaxId find_lca(const std::vector<NormalizedTaxId> &taxids);
+
+  public:
+    /**
+     * Constructs a TaxonomyDB
+     *
+     * @param [input] taxo_tree_filepath path to a "nodes.dmp" file.
+     * @param [input] lookup_table_filepath path to a ".accession2taxid" file.
+     * @param [input] fasta_headers_filepath path to a ".fasta.fai" file.
+     */
+    TaxonomyDB(const std::string &taxo_tree_filepath,
```

Comment:
Regarding the integration tests for this class, one idea can be to use the first 100.000 lines of the virus sequences from GenBank `/cluster/work/grlab/projects/metagenome/data/viruses/virus_database_graph`. There will be around 100 contigs with a running time of 60-90 seconds. What do you think?